Remotion/Data/DomainObjects/Persistence/Rdbms/Parameters/ObjectIDDataParameterDefinition.cs

```diff
@@ -0,0 +1,66 @@
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Data;
+using Remotion.Data.DomainObjects.Persistence.Configuration;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+
+/// <summary>
+/// Can create <see cref="IDbDataParameter"/> instances for <see cref="ObjectID"/> parameter values that refer to <see cref="DomainObject"/>s stored with the same
+/// <see cref="StorageProviderDefinition"/>.
+/// </summary>
+public class ObjectIDDataParameterDefinition : IDataParameterDefinition
+{
+  public IStorageTypeInformation ValueStorageTypeInformation { get; }
+
+  /// <summary>
+  /// Creates a new <see cref="ObjectIDDataParameterDefinition"/> instance.
+  /// </summary>
+  /// <param name="storageTypeInformation">The <see cref="IStorageTypeInformation"/> associated with the <see cref="Type"/> of the <see cref="ObjectID"/>'s <see cref="ObjectID.Value"/>.</param>
+  public ObjectIDDataParameterDefinition (IStorageTypeInformation storageTypeInformation)
+  {
+    ArgumentUtility.CheckNotNull(nameof(storageTypeInformation), storageTypeInformation);
+
+    ValueStorageTypeInformation = storageTypeInformation;
+  }
+
+  /// <inheritdoc />
+  public object GetParameterValue (object? parameterObject)
+  {
+    var objectID = ArgumentUtility.CheckType<ObjectID>(nameof(parameterObject), parameterObject);
+    var value = objectID?.Value;
+    if (value != null && !ValueStorageTypeInformation.DotNetType.IsInstanceOfType(value))
+      throw ArgumentUtility.CreateArgumentTypeException("parameterObject.Value", value.GetType(), ValueStorageTypeInformation.DotNetType);
```

Comment:
Checktype does not support nullability in the way we need here, so we need to check for that extra.

Reply:
Am not sure if there's not an expectation mismatch. We need a check that ignores the test of value is null and otherwise, tests if if Value is assignable to the expected type.

This should be covered by 
public static object? CheckType ([InvokerParameterName] string argumentName, [NoEnumeration] object? actualValue, Type expectedType)

I see that the question is whether we want to handle StorageTypeInformation.DotNetType being an Int32, but the value being null and if this should actually be a DotNetType Nullable<Int32> instead.  From reading the code (mistake is possible) I think that DotNetType would be Nullable<int32> in such cases. Or it would be a correct argumentexception otherwise.

Please check/verify.